Add support for Windows Terminal

codeofdusk opened the pull request:
<!--
Please fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/wiki/Github-pull-request-template-explanation-and-examples
-->

### Link to issue number:
Closes #10305. Supersedes #10716 for now.

### Summary of the issue:
The new Windows Terminal does not support automatic readout, suppression of password characters, and other terminal niceties.

Since newer Windows Console builds correctly set the end of text ranges as exclusive, NVDA's current UIA console workarounds produce incorrect behaviour in these builds.

### Description of how this pull request fixes the issue:
* Like #10716, splits the `consoleUIATextInfo` workarounds into a subclass. This will be useful for Windows 10 version 21H1, where making UIA default in consoles will be strongly worth considering.
* Adds a new `winConsoleUIA.WinTerminalUIA` class and attaches it to Windows Terminal controls.
* Re-introduces support for `IUIAutomationTextPattern::GetVisibleRanges` for consoles returning one contiguous range (see microsoft/terminal#4495).
* Factors out the password/doubled character suppression into a `NVDAObjects.behaviors.EnhancedTermTypedCharSupport` class, which `KeyboardHandlerBasedTypedCharSupport` now subclasses.
* Adds new logic to properly choose the `textInfo` implementation to use based on the number of visible ranges presented by the console (newer builds present just one range).

### Testing performed:
Tested that Windows Terminal controls correctly receive the `WinTerminalUIA` type and that consoles work as expected (including keyboard support enhancements).

Tested that in newer Windoes Console (built from the terminal `master` branch), bounding works as expected even after the window is maximized and a large amount of text is written.

Tested that automatic readout, password suppression, and typed character reporting are functional in the Windows Terminal build in microsoft/terminal#2447.

Tested that the correct `textInfo` implementation is chosen for the included console on Windows 10 version 1909 Vs. the latest open console from Microsoft.

### Known issues with pull request:
None.

### Change log entry:
== new features ==
- Added support for Windows Terminal. (#10305)

codeofdusk committed:
Bring over ConsoleUIATextInfo refactoring from #10716.

codeofdusk committed:
Add WinTerminalUIA

codeofdusk committed:
Style.

codeofdusk commented:
Cc @leonardder @michaelDCurran

lukaszgo1 commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -48,6 +47,56 @@ def __init__(self, obj, position, _rangeObj=None):
+	def move(self, unit, direction, endPoint=None):
+		oldInfo = None
+		if self.basePosition != textInfos.POSITION_CARET:
+			# Insure we haven't gone beyond the visible text.
```
Insure -> Ensure

codeofdusk committed:
Review actions.

codeofdusk committed:
microsoft/terminal#4495: support fixed GetVisibleRanges.

codeofdusk committed:
KeyboardHandlerBasedTypedCharSupport -> EnhancedTermTypedCharSupport, add _shouldUseToUnicodeEx flag and set False for WinTerminalUIA.

codeofdusk commented:
@leonardder @michaelDCurran Could you please have another look? I think this is now ready for merge, and will become functional with microsoft/terminal#4826.

codeofdusk commented:
The PR description has also been updated with additional changes and testing notes.

codeofdusk committed:
Update doc comments.

codeofdusk committed:
Update copyrights.

codeofdusk commented:
Cc @feerrenrut and @michaelDCurran Could this PR please be reviewed?

feerrenrut commented:
I haven't yet looked through the changes carefully, my initial impression is concern for regressions. Anything you can do here to reduce the chance of regression (particularly in windows console) will help us to get confidence in this PR. 

We are planning to try to clear the backlog of PR's soon, we will initially focus on bug fixes and small PR's.

codeofdusk commented:
> I haven't yet looked through the changes carefully, my initial impression is concern for regressions. Anything you can do here to reduce the chance of regression (particularly in windows console) will help us to get confidence in this PR.
> 
> We are planning to try to clear the backlog of PR's soon, we will initially focus on bug fixes and small PR's.

Chance of regressions seems fairly low, as the new terminal and old console functionality are in separate subclasses (`WinConsoleUIA` and `WinTerminalUIA`). However, this PR does lay the groundwork for improved UIA console support in Windows 10 2104, which may (likely will) be good enough to use as default when new conhost (including microsoft/terminal#4018) ships.

dingpengyu commented:
hello 
Is pr currently ready to merge?
thanks

codeofdusk commented:
> hello
> Is pr currently ready to merge?
> thanks

Yes, once reviewed by @feerrenrut or @michaelDCurran.

codeofdusk commented:
Oops, ignore 6c9f010 (patch for changes in a different PR).

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -13,43 +13,116 @@
+		# microsoft/terminal#4495: In newer consoles,
+		# IUIAutomationTextRange::getVisibleRanges returns a reliable contiguous range.
+		visiRange = None
+		try:
```
How does this code know whether the fetched visible range is appropriate on older versions of Windows 10?

codeofdusk replied:
If the console returns more than one range then it's an older version of Windows, so visible ranges are unreliable. Ideally we'd always use visible ranges on `ConsoleUIATextInfo` instances and always rangeFromPoint on `ConsoleUIATextInfoEndInclusive` instances, but I'm not sure how to cleanly do this.

leonardder replied:
That's exactly what I wanted to suggest.
You could always split out stuff to one or more helper methods that are called by __init__.

codeofdusk replied:
> That's exactly what I wanted to suggest.
> You could always split out stuff to one or more helper methods that are called by **init**.

Where would the separation into separate methods most make sense here? In case of `rangeFromPoint` we could write something that returns the correct `_RangeObj` based on the passed-in position for when `super` is called,but for visible ranges we need to call `collapse` *after* `super` is called.

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -13,43 +13,116 @@
+class consoleUIATextInfoEndInclusive(consoleUIATextInfo):
 	def collapse(self, end=False):
-		"""Works around a UIA bug on Windows 10 1803 and later."""
+		"""Works around a UIA bug on Windows 10 versions before 2104."""
```
```suggestion
		"""Works around a UIA bug on Windows 10 versions before 2004."""
```

codeofdusk replied:
Still broken on 2004. The fix will be in a 21xx build according to Microsoft.

leonardder replied:
AH, that clarifies why you're using 2104 all the time :). Still we aren't sure whether 2104 will be the name Microsoft will use, so may be post 2004 is better.

codeofdusk replied:
> AH, that clarifies why you're using 2104 all the time :). Still we aren't sure whether 2104 will be the name Microsoft will use, so may be post 2004 is better.

Wouldn't that possibly be confusing for version 2010? (i.e. that's post-2004 and won't have the fix)

leonardder replied:
True. Still, i t's only a promise from MS that they will fix it in 2104.
May be 21xx?

leonardder replied:
Let's go with 21H1

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -62,15 +135,65 @@ def collapse(self, end=False):
-			oldInfo = self.copy()
-		if unit == textInfos.UNIT_WORD and direction != 0:
+	def compareEndPoints(self, other, which):
+		"""Works around a UIA bug on Windows 10 versions before 2104."""
```
```suggestion
		"""Works around a UIA bug on Windows 10 versions before 2004."""
```

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -62,15 +135,65 @@ def collapse(self, end=False):
+
+	def setEndPoint(self, other, which):
+		"""Override of L{textInfos.TextInfo.setEndPoint}.
+		Works around a UIA bug on Windows 10 versions before 2104 that means we can
```
```suggestion
		Works around a UIA bug on Windows 10 versions before 2004 that means we can
```

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -62,15 +135,65 @@ def collapse(self, end=False):
+
+	def _move(self, unit, direction, endPoint=None):
+		if unit == textInfos.UNIT_WORD and direction != 0:
+			# On Windows 10 versions before 2104, UIA doesn't implement word
```
```suggestion
			# On Windows 10 versions before 2004, UIA doesn't implement word
```

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -13,43 +13,116 @@
+		return super(consoleUIATextInfo, self).move(unit, direction, endPoint)
+
+	def __ne__(self, other):
+		"""Support more accurate caret move detection."""
```
It's not clear to me how overriding __ne__ can help in caret movement detection.

codeofdusk replied:
This is not new code (is currently on `ConsoleUIATextInfo`). It prevents the last character of the prompt from being read out.

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -258,26 +291,27 @@ def _getWordOffsetsInThisLine(self, offset, lineInfo):
-		"""Support more accurate caret move detection."""
-		return not self == other
+	def _isCollapsed(self):
+		"""Works around a UIA bug on Windows 10 versions before 2104 that means we
```
```suggestion
		"""Works around a UIA bug on Windows 10 versions before 2004 that means we
```

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -13,43 +13,116 @@
+			return res
+
 
+class consoleUIATextInfoEndInclusive(consoleUIATextInfo):
```
I have the feeling that this does more than being end inclusive, may be something like this

```suggestion
class consoleUIATextInfoPre2004(consoleUIATextInfo):
```

leonardder commented on source/NVDAObjects/behaviors.py:
```diff
@@ -400,12 +400,16 @@ class KeyboardHandlerBasedTypedCharSupport(Terminal):
 	#: Timely and reliable textChange events are required
 	#: to support password suppression.
 	_supportsTextChange = True
+	#: Whether to use ToUnicodeEx for typed character events.
```
What's the benefit of this class when this is False?

codeofdusk replied:
Catches typed characters that aren't written to the screen (such as passwords), handles tab completion, etc.

leonardder replied:
AH I see, so it gets the keybord input by other means.

leonardder commented on source/NVDAObjects/behaviors.py:
```diff
@@ -390,8 +390,8 @@ def _get_caretMovementDetectionUsesEvents(self):
 
-class KeyboardHandlerBasedTypedCharSupport(Terminal):
-	"""A Terminal object that also provides typed character support for
+class EnhancedTermTypedCharSupport(Terminal):
```
Is it really necessary to rename this? It breaks compatibility.

codeofdusk replied:
Compatibility with what?

leonardder replied:
Backwards compatibility with older versions of NVDA. It is unlikely but possible that add-ons use this class.

codeofdusk replied:
> Backwards compatibility with older versions of NVDA. It is unlikely but possible that add-ons use this class.

If having a class named `KeyboardHandlerBasedTypedCharSupport` that doesn't actually use `keyboardHandler` in some cases isn't that confusing, I can revert the rename.

leonardder replied:
May be you could re-add KeyboardHandlerBasedTypedCharSupport as a subclass of EnhancedTermTypedCharSupport, it can just do nothing. This way, you could also remove the _shouldUseToUnicodeEx attribute, as the isinstance check should be sufficient.

LeonarddeR reviewed (changes requested):
Have you tested this on insider builds of Windows 2004?

josephsl commented:
Hi, or better yet, it might be best to call it 21H1, as we don’t know if it’s going to be called 2103 or 2104 – won’t find out until December at the earliest. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Monday, March 30, 2020 11:58 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [nvaccess/nvda] Add support for Windows Terminal (#10784)

 

@leonardder commented on this pull request.

  _____  

In source/NVDAObjects/UIA/winConsoleUIA.py <https://github.com/nvaccess/nvda/pull/10784#discussion_r400682925> :

>       def collapse(self, end=False):
-              """Works around a UIA bug on Windows 10 1803 and later."""
+              """Works around a UIA bug on Windows 10 versions before 2104."""

True. Still, i t's only a promise from MS that they will fix it in 2104.
May be 21xx?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/10784#discussion_r400682925> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4AXEA6PZRKLVJBS4GPMEDRKGIAFANCNFSM4KUZFIWQ> .

codeofdusk committed:
Review actions.

codeofdusk committed:
Factor out bounding into separate method.

codeofdusk committed:
Re-introduce KeyboardHandlerBasedTypedCharSupport.

codeofdusk commented:
@leonardder I think all requested changes have now been made.

codeofdusk committed:
Fix UIA console support for newer Windows Console builds.

codeofdusk committed:
Revert copyright header changes.

codeofdusk commented:
I've rebased this PR on master as it's been open for a while.

@leonardder Could you please have another look?

codeofdusk committed:
Revert another no longer necessary copyright header change.

codeofdusk committed:
Fix comments.

dingpengyu commented:
hi @codeofdusk 
Is pr basically complete now?
thanks

codeofdusk commented:
> hi @codeofdusk
> Is pr basically complete now?
> thanks

Yes, awaiting review from @leonardder and @michaelDCurran.

codeofdusk committed:
Update docstring.

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -13,19 +13,103 @@
+		and whether the textInfo should be collapsed after instantiation."""
+		# microsoft/terminal#4495: In newer consoles,
+		# IUIAutomationTextRange::getVisibleRanges returns a reliable contiguous range.
+		_rangeObj = obj.UIATextPattern.GetVisibleRanges().GetElement(0)
```
Could you please put this in a try/except block and raise a RuntimeError when this fails? It should not fail, but better safe than sorry. COMErrors raised directly or indirectly from TextInfo.__init__ usually aren't caught by callers of makeTextInfo.

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -13,19 +13,103 @@
+			return res
+
+
+class consoleUIATextInfoPre21H1(consoleUIATextInfo):
```
Please add a class doc string

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -62,15 +146,65 @@ def collapse(self, end=False):
-		if unit == textInfos.UNIT_WORD and direction != 0:
+	def compareEndPoints(self, other, which):
+		"""Works around a UIA bug on Windows 10 versions before 21H1."""
+		# Even when a console textRange's start and end have been moved to the
```
Could you add these comments to the doc string instead?

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -297,12 +342,19 @@ def _get_windowThreadID(self):
+		"Returns whether this is a newer version of Windows Console with an improved UIA implementation."
+		# microsoft/terminal#4495: In newer consoles,
+		# IUIAutomationTextRange::getVisibleRanges returns one visible range.
+		return self.UIATextPattern.GetVisibleRanges().length == 1
```
Are you absolutely sure that pre21H1 will never return only one visible range?

codeofdusk replied:
Only in a situation where the console displays *exactly* one line, which seems impossible.

leonardder commented on source/NVDAObjects/behaviors.py:
```diff
@@ -490,6 +484,20 @@ def _findNonBlankIndices(self, lines):
+	even when a character is not written to the console (e.g. in read only console applications).
+	This approach is an alternative to monitoring the console output for
+	characters close to the caret, or injecting in-process with NVDAHelper.
+	This class relies on the toUnicodeEx Windows function, and in particular
```
```suggestion
	This class does not implement any specific functionality by itself.
	Rather, it instructs keyboardHandler to use the toUnicodeEx Windows function, in particular
```

LeonarddeR reviewed (changes requested):
Yay, this shortens the diff significantly, that's good!

codeofdusk committed:
Review actions.

codeofdusk committed:
Update source/NVDAObjects/behaviors.py

Co-Authored-By: Leonard de Ruijter <leonardder@users.noreply.github.com>

codeofdusk committed:
Add class docstring.

codeofdusk commented:
@leonardder @feerrenrut Could you please have another look?

codeofdusk committed:
Merge branch 'master' into t10305

LeonarddeR reviewed (approved):
Thanks @codeofdusk

michaelDCurran commented:
Try build: https://ci.appveyor.com/api/buildjobs/milyub46w9dmxg7l/artifacts/output%2Fnvda_snapshot_try-t10305-19897%2C6ff9867f.exe

michaelDCurran commented:
I shall try to review this on Monday.

codeofdusk commented:
Perhaps in a separate PR, we may need to rethink how and when we clear the typed word buffer for new text lines in `EnhancedTermTypedCharSupport._calculateNewText`.

On my system:

1. Enable "speak typed words".
2. Open Open Console (the 21H1 conhost).
3. Run `git log`.
4. Press q to close the pager.
5. Type `git` and press space.

NVDA says "qgit", not "git".

For reference, the 21H1 console (`OpenConsole.exe`) and Windows Terminal (`WindowsTerminal.exe`) as of commit microsoft/terminal@4f8acb4 can be found [here](https://github.com/nvaccess/nvda/files/4432945/CascadiaPackage_4f8acb4_x64.zip). On my system, I've replaced the default conhost with this version (see microsoft/terminal#1817).

codeofdusk committed:
Optimization: clear character buffers when reporting new lines.

This saves us from scanning the old and new lines twice in _calculateNewText (once for diffing and once for finding nonblank indices), and fixes typed word reporting for 21H1 console.

This new approach does not resolve #10942.

codeofdusk committed:
Update inaccurate information in user guide.

michaelDCurran commented:
@codeofdusk  I'm happy with this. Are you ready for me to merge this?

codeofdusk commented:
Yes – Future changes, if necessary, will be made in subsequent PRs.
